hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/placement/algorithms/SCMContainerPlacementRackScatter.java

```diff
@@ -290,24 +314,45 @@ protected List<DatanodeDetails> chooseDatanodesInternal(
       unavailableNodes.addAll(excludedNodes);
     }
 
+    LOG.debug("Available racks excluding racks with used nodes: {}.", racks);
     if (racks.size() < additionalRacksRequired) {
       String reason = "Number of existing racks: " + racks.size()
               + "is less than additional required number of racks to choose: "
               + additionalRacksRequired + " do not match.";
       LOG.warn("Placement policy cannot choose the enough racks. {}"
                       + "Total number of Required Racks: {} Used Racks Count:" +
-                      " {}, Required Nodes count: {}",
+                      " {}, Required Nodes count: {}, fallback: {}.",
               reason, numberOfRacksRequired, usedRacksCntMap.size(),
-              nodesRequired);
-      throw new SCMException(reason,
-              FAILED_TO_FIND_SUITABLE_NODE);
+              nodesRequired, fallback);
+
+      /*
+      The number of racks which aren't already used by existing replicas
+      ("racks") and hence are available for new replicas are less than the
+      additional number of racks that replicas need to be on
+      ("additionalRacksRequired").
+      */
+      if (!fallback) {
+        throw new SCMException(reason,
+            FAILED_TO_FIND_SUITABLE_NODE);
+      }
     }
 
+    /*
+    If fallback is false, and we reach here, then nodesRequired is greater
+    than or equal to additionalRacksRequired. So we need to ask for
+    nodesRequired number of nodes. If fallback is true, then nodesRequired
+    could be less than additionalRacksRequired. We still need to ask for
+    nodesRequired number of more nodes, even if that won't satisfy the
+    requirement of being on additionalRacksRequired more racks.
+
+    So, in both cases, we want to ask for nodesRequired number of nodes.
+    */
+    int maxOuterLoopIterations = Math.min(nodesRequired, maxReplicasPerRack);
     Set<DatanodeDetails> chosenNodes = new LinkedHashSet<>(
         chooseNodesFromRacks(racks, unavailableNodes,
-            mutableFavoredNodes, additionalRacksRequired,
-            metadataSizeRequired, dataSizeRequired, 1,
-            usedRacksCntMap, maxReplicasPerRack));
+            mutableFavoredNodes, nodesRequired, metadataSizeRequired,
+            dataSizeRequired, maxOuterLoopIterations, usedRacksCntMap,
```

Comment:
I changed this because it's possible to have a situation where `nodesRequired` is less than `additionalRacksRequired` (check the third case in my linked document). If we set it to `nodesRequired`, the policy still prefers finding a new rack rather than a rack that already has a used node.

Reply:
I believe in that case we should just check if the misreplication count is going down or remaining same in the case fallback is enabled. We shouldn't change this parameter.

Reply:
> check if the misreplication count is going down or remaining same in the case fallback is enabled

I believe we're already doing that here:
```
    List<DatanodeDetails> newPlacement =
        new ArrayList<>(usedNodes.size() + result.size());
    newPlacement.addAll(usedNodes);
    newPlacement.addAll(chosenNodes);
    ContainerPlacementStatus placementStatus =
        validateContainerPlacement(newPlacement, requiredReplicationFactor);
    if (!placementStatus.isPolicySatisfied()) {
      ContainerPlacementStatus initialPlacementStatus =
          validateContainerPlacement(usedNodes, requiredReplicationFactor);
      if (initialPlacementStatus.misReplicationCount()
              < placementStatus.misReplicationCount()) {
        String errorMsg = "ContainerPlacementPolicy not met. Misreplication" +
                " Reason: " + placementStatus.misReplicatedReason() +
                " Initial Used nodes mis-replication Count: " +
                initialPlacementStatus.misReplicationCount() +
                " Used nodes + Chosen nodes mis-replication Count: " +
                placementStatus.misReplicationCount();
        throw new SCMException(errorMsg, FAILED_TO_FIND_SUITABLE_NODE);
      }
    }
```

> We shouldn't change this parameter.

My understanding is that we should look for `additionalRacksRequired` if fallback is false, and for `nodesRequired` if fallback is true. Now, if fallback is false, then `nodesRequired` must be greater than `additionalRacksRequired` or we'll throw an exception earlier in the method:
```    if (nodesRequired < additionalRacksRequired) {
      String reason = "Required nodes size: " + nodesRequired
              + " is less than required number of racks to choose: "
              + additionalRacksRequired + ".";
      LOG.warn("Placement policy cannot choose the enough racks. {}"
                      + "Total number of Required Racks: {} Used Racks Count:" +
                      " {}, Required Nodes count: {}, fallback: {}.",
              reason, numberOfRacksRequired, usedRacksCntMap.size(),
              nodesRequired, fallback);

      /*
      The number of additional Datanodes ("nodesRequired") that were asked for
      are less than the additional number of racks ("additionalRacksRequired")
      that container replicas need to be on. This means that returning
      nodesRequired number of nodes will not satisfy the rack scatter policy.
      If fallback is false, throw an exception. Else, continue to find
      nodesRequired number of nodes.
       */
      if (!fallback) {
        throw new SCMException(reason,
            FAILED_TO_FIND_SUITABLE_NODE);
      }
    }
```

So this means that for both values of fallback, the argument can just be `nodesRequired`. Anyway, `chooseNodesFromRacks` should still try to spread it across as many racks as possible, with excluded and used racks sorted at the end of the list. What do you think?